File: client/src/utils/validatePassword.js
export function validatePassword(password) {
  const upperRegex = /[A-Z]/;
  const lowerRegex = /[a-z]/;
  const numberRegex = /[0-9]/;
  const specialRegex = /[!@#\$%\^&\*.,?\[\]\(\)\{\}]/;

  // password too short
  if (password.length < 8) {
    return { valid: false, message: 'password too short' };
  }

  // no uppercase
  if (upperRegex.test(password) === false) {
    return { valid: false, message: 'must contain an uppercase' };
  }

  // no lowercase
  if (lowerRegex.test(password) === false) { 
    return { valid: false, message: 'must contain a lowercase' };
  }

  // no number
  if (numberRegex.test(password) === false) {
    return { valid: false, message: 'must contain a number' };
  }

  // no special character
  if (specialRegex.test(password) === false) {
    return { valid: false, message: 'must contain a special character' };
  }

  return { valid: true, message: 'password is valid' };
}

File: server/core/utils/validate_password.py
import re

def validate_password(password):
    UPPER_REGEX = re.compile(r'[A-Z]')
    LOWER_REGEX = re.compile(r'[a-z]')
    DIGIT_REGEX = re.compile(r'[0-9]')
    SPECIAL_REGEX = re.compile(r'[!@#\$%\^&\*.,?\[\]\(\)\{\}]')

    # password too short
    if len(password) < 8:
        return False
    
    # no uppercase letter
    if not UPPER_REGEX.search(password):
        return False
    
    # no lowercase letter
    if not LOWER_REGEX.search(password):
        return False
    
    # no digit
    if not DIGIT_REGEX.search(password):
        return False
    
    # no special character
    if not SPECIAL_REGEX.search(password):
        return False
    
    return True